### ISSUE TITLE
Different behavior of LSTM and LSTMCell implementation 

### ISSUE THREAD

**Issue by DeepInEvil** [pytorch/pytorch#8242]
## Issue description
I was testing the difference between LSTM and LSTMCell implementations, ideally for same input they should have same outputs, but the outputs are different, looks like something fishy is going on.
## Code example
import torch
from torch import nn
from torch.autograd import Variable
print torch.__version__
torch.manual_seed(666)
rnn = nn.LSTM(10, 20, 1)
input = Variable(torch.randn(5, 3, 10))
h0 = Variable(torch.zeros(1, 3, 20))
c0 = Variable(torch.zeros(1, 3, 20))
output, (hn, cn) = rnn(input, (h0, c0))

print ("From LSTM")
print hn[0]

rnncell = nn.LSTMCell(10, 20)
hx = Variable(torch.zeros(3, 20))
cx = Variable(torch.zeros(3, 20))
out = []
for i in range(5):
    hx, cx = rnncell(input[i], (hx, cx))
    out.append(hx)

print ("from LSTM Cell")
print hx
## System Info

- PyTorch or Caffe2: pytorch
- How you installed PyTorch (conda, pip, source): pip
- Build command you used (if compiling from source):
- OS: ubuntu
- PyTorch version: 0.3.1 
- Python version: 2.7.12
- CUDA/cuDNN version: NA
- GPU models and configuration: NA
- GCC version (if compiling from source):
- CMake version: NA
- Versions of any other relevant libraries:

**Comment by SsnL** [comment]
hmm aren't they using different weights in your example? so different results is expected?

**Comment by DeepInEvil** [comment]
Hi, you are right, but how would you explain the difference in performances?

**Comment by SsnL** [comment]
Perf should be same. Do you have a minimal example script to show the perf difference?

**Comment by zou3519** [comment]
@DeepInEvil if you mean "why is LSTMCell slower than LSTM", there's a fused LSTM implementation in C when you run LSTM but not LSTMCell.

**Comment by SsnL** [comment]
@zou3519 I think @DeepInEvil is sarying perf as in accuracy etc., not time.

**Comment by abhinavkashyap** [comment]
Is there any progress on this? Even I find that the difference in my performance metrics (accuracy) between using `LSTM` and `LSTMCell` to be huge.

**Comment by zou3519** [comment]
@abhinavkashyap do you find using double vs float to help? I'm wondering if this is a numerical accuracy problem

**Comment by shibizhao** [comment]
Is there any progress on this?

**Comment by ndminhkhoi46** [comment]
@DeepInEvil  Did you make sure their initialization is the same? Different initialization would lead to different result due to the complexity of the model. 
@zou3519 A source for the differences in implementation in LSTM and LSTMCell would help answering the differences in perf.
